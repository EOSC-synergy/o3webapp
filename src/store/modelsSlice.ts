import { createSlice } from '@reduxjs/toolkit';
import {
    PROCESS_SV_WITH_PERCENTILE,
    STATISTICAL_VALUES,
    STATISTICAL_VALUES_LIST,
} from 'utils/constants';
import { HYDRATE } from 'next-redux-wrapper';
import { pick } from 'lodash';

type Optional<T, K extends keyof T> = Pick<Partial<T>, K> & Omit<T, K>;

type ModelData = {
    // single model
    color: string | null; // if not set it defaults to standard value from api
    isVisible: boolean; // show/hide individual models from a group
    // init statistical values dynamically with for-loop
} & Optional<
    Record<PROCESS_SV_WITH_PERCENTILE, boolean>,
    'lowerPercentile' | 'upperPercentile' | 'stdMean'
>;
/**
 * This object serves as a template, whenever a new model is added to a group this object describes
 * the default settings of the freshly added model.
 *
 * @category ModelSlice
 * @constant {Object}
 * @default { color: null, isVisible: true }
 */
export const MODEL_DATA_TEMPLATE: ModelData = {
    // single model
    color: null, // if not set it defaults to standard value from api
    isVisible: true, // show/hide individual models from a group
    ...(PROCESS_SV_WITH_PERCENTILE.reduce((p, c) => {
        return { ...p, [c]: true };
    }, {}) as Record<PROCESS_SV_WITH_PERCENTILE, boolean>),
};

export type ModelGroup = {
    name: string;
    models: Record<string, ModelData>;
    isVisible: boolean;
    visibleSV: Record<STATISTICAL_VALUES, boolean>;
};

/**
 * This object serves as a template to a Model Group and defines major fields that have to be filled
 * if a new model group is added.
 *
 * @category ModelSlice
 * @constant {Object}
 * @default { name: "",
 *     models: {},
 *     isVisible: true,
 *     visibleSV: {
 *     }}
 */
const MODEL_GROUP_TEMPLATE: ModelGroup = {
    name: '',
    models: {}, // models is lookup table
    isVisible: true, // show/hide complete group
    visibleSV: STATISTICAL_VALUES_LIST.reduce((p, c) => {
        return { ...p, [c]: true };
    }, {}) as Record<STATISTICAL_VALUES, boolean>,
};

export type ModelId = string;

export type ModelState = {
    idCounter: number;
    modelGroups: Record<ModelId, ModelGroup>;
};

export type GlobalModelState = {
    models: ModelState;
};

/**
 * The initial state of the modelSlice defines the data structure in the store. Each plot has its
 * own data i.e. have separate model(groups).
 *
 * IF you change this initial state you have to adapt the first test in the corresponding test file,
 * that tests the initial state.
 *
 * @category ModelSlice
 * @constant {object}
 * @default { idCounter: 0, modelGroups: {} }
 */
const initialState: ModelState = {
    idCounter: 0,
    // currently active plot
    modelGroups: {},
};

type Payload<T> = {
    payload: T;
};

/**
 * The modelSlice is generated by the redux toolkit. The reducers are defined here and the
 * corresponding actions are auto-generated.
 *
 * @category ModelSlice
 * @constant {Object}
 */
const modelsSlice = createSlice({
    name: 'models',
    initialState: initialState,
    reducers: {
        /**
         * This reducer accepts an action object returned from setModelsOfModelGroup()
         *
         *     e.g.dispatch(
         *         setModelsOfModelGroup({
         *             groupId: 0,
         *             modelList: ['CCMI-1_ACCESS_ACCESS-CCM-refC2', 'CCMI-1_CCCma_CMAM-refC2'],
         *         })
         *     );
         *
         * And calculates the new state based on the action and the action data given in
         * action.payload.
         *
         * This method provides a convenient interface for the AddModelGroupModal by allowing to
         * dispatch a groupId with the required models. If the group already exists the
         * corresponding data is updated otherwise the reducer TAKES CARE of creating a group.
         *
         * @example
         *     dispatch(
         *         setModelsOfModelGroup({
         *             groupId: 0,
         *             modelList: ['CCMI-1_ACCESS_ACCESS-CCM-refC2', 'CCMI-1_CCCma_CMAM-refC2'],
         *         })
         *     );
         *
         * @param {object} state The current store state of: state/models
         * @param {object} action Accepts the action returned from updateModelGroup()
         * @param {object} action.payload The payload is an object containing the given data
         * @param {int} action.payload.groupId The name of the group to set
         * @param {string} action.payload.groupName The name of the group
         * @param {string} action.payload.modelList The list of models the group should have
         */
        setModelsOfModelGroup(
            state: ModelState,
            {
                payload: { groupId, groupName, modelList },
            }: Payload<{ groupId?: number; groupName: string; modelList: string[] }>
        ) {
            // set model group
            if (groupId != null && groupId in state.modelGroups) {
                const selectedModelGroup = state.modelGroups[groupId];
                selectedModelGroup.name = groupName;
                // remove unwanted
                selectedModelGroup.models = pick(selectedModelGroup.models, modelList);

                // add new ones
                for (const model of modelList) {
                    if (!(model in selectedModelGroup.models)) {
                        // initialize with default settings
                        selectedModelGroup.models[model] = { ...MODEL_DATA_TEMPLATE };
                    }
                }
            } else {
                // create new group
                state.modelGroups[state.idCounter++] = {
                    ...MODEL_GROUP_TEMPLATE,
                    name: groupName,
                    models: modelList.reduce((acc, model) => {
                        acc[model] = { ...MODEL_DATA_TEMPLATE };
                        return acc;
                    }, {} as Record<string, ModelData>),
                };
            }
            // change name either way
        },

        /**
         * This reducer accepts an action object returned from deleteModelGroup()
         *
         *     e.g.dispatch(deleteModelGroup({ 0: 'refC2' }));
         *
         * And calculates the new state based on the action and the action data given in
         * action.payload.
         *
         * This method provides a convenient interface for the AddModelGroupModal by allowing to
         * dispatch a groupId with the required models. If the group already exists the
         * corresponding data is updated otherwise the reducer takes care of creating a group.
         *
         * @example
         *     dispatch(deleteModelGroup({ 0: 'refC2' }));
         *
         * @param state The current store state of: state/models
         * @param groupId The name of the group that should be deleted
         */
        deleteModelGroup(
            state: ModelState,
            { payload: { groupId } }: Payload<{ groupId: number | string }>
        ) {
            if (!(groupId in state.modelGroups)) {
                // no group with this name in store
                throw new Error(`tried to delete "${groupId}" which is not a valid group`);
            }

            delete state.modelGroups[groupId]; // delete from lookup table
        },

        /**
         * This reducer accepts an action object returned from updatePropertiesOfModelGroup()
         *
         *     e.g.dispatch(setStatisticalValueForGroup({ groupID: 42, data: bigObject }));
         *
         * This method provides an interface to update the properties of an existing model group.
         * The properties are whether the model is included in the statistical value(s) and whether
         * the model is visible.
         *
         * @example
         *     dispatch(setStatisticalValueForGroup({ groupID: 42, data: bigObject }));
         *
         * @param state The current store state of: state/models
         * @param groupId The name of the group whose model properties should be updated
         * @param data Holds the information that should be updated
         */
        updatePropertiesOfModelGroup(
            state: ModelState,
            {
                payload: { groupId, data },
            }: Payload<{ groupId: number | string; data: ModelGroup['models'] }>
        ) {
            if (!(groupId in state.modelGroups)) {
                // no group with this name in store
                throw new Error(`tried to access "${groupId}" which is not a valid group`);
            }

            const modelGroup = state.modelGroups[groupId];
            for (const model in modelGroup.models) {
                modelGroup.models[model] = data[model];
            }
        },

        /**
         * This reducer accepts an action object returned from setStatisticalValueForGroup() e.g.
         * dispatch(setStatisticalValueForGroup( {groupID: 0, svType: STATISTICAL_VALUES.median,
         * isIncluded: true} )); and calculates the new state based on the action and the action
         * data given in action.payload.
         *
         * In this case for a given group is set whether the given statistical values (SV) should be
         * displayed.
         *
         * @example
         *     dispatch(
         *         setStatisticalValueForGroup({
         *             groupID: 0,
         *             svType: STATISTICAL_VALUES.median,
         *             isIncluded: true,
         *         })
         *     );
         *
         * @param {object} state The current store state of: state/models
         * @param {object} action Accepts the action returned from updateModelGroup()
         * @param {object} action.payload The payload is an object containing the given data
         * @param {int} action.payload.groupId A string specifying the group
         * @param {string} action.payload.svType The SV as a string
         * @param {boolean} action.payload.isIncluded Should the SV be displayed for the given group
         */
        setStatisticalValueForGroup(
            state: ModelState,
            {
                payload: { groupId, svType, isIncluded },
            }: Payload<{
                groupId: number | string;
                svType: STATISTICAL_VALUES;
                isIncluded: boolean;
            }>
        ) {
            if (!STATISTICAL_VALUES_LIST.includes(svType)) {
                // svType doesn't represent a valid statistical value
                throw new Error(
                    `tried to set statistical value "${svType}" that is not a valid statistical value (${STATISTICAL_VALUES_LIST.join(
                        '|'
                    )})`
                );
            }
            if (!(groupId in state.modelGroups)) {
                // no group with this name in store
                throw new Error(`tried to access "${groupId}" which is not a valid group`);
            }

            state.modelGroups[groupId].visibleSV[svType] = isIncluded;
        },

        /**
         * This reducer accepts an action object returned from setVisibilityForGroup() e.g.
         * dispatch(setVisibilityForGroup( {groupID: 0, isVisible: true} )); and calculates the new
         * state based on the action and the action data given in action.payload.
         *
         * In this case for a given group is set whether it should be visible or not.
         *
         * @example
         *     dispatch(setVisibilityForGroup({ groupID: 0, isVisible: true }));
         *
         * @param state The current store state of: state/models
         * @param groupId A string specifying the group
         * @param isVisible Should the group be visible
         * @throws An Error if the provided groupId is not valid
         */
        setVisibilityForGroup(
            state: ModelState,
            {
                payload: { groupId, isVisible },
            }: Payload<{ groupId: number | string; isVisible: boolean }>
        ) {
            if (!(groupId in state.modelGroups)) {
                throw new Error(`tried to access "${groupId}" which is not a valid group`);
            }

            state.modelGroups[groupId].isVisible = isVisible;
        },
    },
    extraReducers: {
        [HYDRATE]: (state: ModelState, action) => {
            console.log('HYDRATE', state, action.payload);
            return {
                ...state,
                ...action.payload.subject,
            };
        },
    },
});

/**
 * The here listed actions are exported and serve as an interface for the view (our React
 * components).
 *
 * @category ModelSlice
 * @constant {Object}
 */
export const {
    setModelsOfModelGroup,
    deleteModelGroup,
    updatePropertiesOfModelGroup,
    setStatisticalValueForGroup,
    setVisibilityForGroup,
} = modelsSlice.actions;

/**
 * The reducer combining all reducers defined in the plot slice. This has to be included in the
 * redux store, otherwise dispatching the above defined actions wouldn't trigger state updates.
 *
 * @category ModelSlice
 */
export default modelsSlice.reducer;
/**
 * This selector allows components to select the models of a given group (specified by ID)
 *
 * @category ModelSlice
 * @function
 * @param {object} state The global redux state
 * @param {int} groupId The group id specifies which group should be retrieved
 * @returns An array containing all models currently in the given group
 */
export const selectModelsOfGroup = (state: GlobalModelState, groupId: number) =>
    Object.keys(state.models.modelGroups[groupId].models);
/**
 * This selector allows components to select the model data of a given group (specified by ID)
 *
 * @category ModelSlice
 * @function
 * @param {object} state The global redux state
 * @param {int} groupId The group id specifies which group data should be retrieved
 * @returns An object containing all the data for each model present in the group. The model names
 *   are the keys.
 */
export const selectModelDataOfGroup = (state: GlobalModelState, groupId?: number) => {
    if (groupId == null) {
        return {};
    }
    if (typeof state.models.modelGroups[groupId] !== 'undefined') {
        return state.models.modelGroups[groupId].models;
    }
    return {};
};
/**
 * This selector allows components to select the name of a given group (specified by ID)
 *
 * @category ModelSlice
 * @function
 * @param {object} state The global redux state
 * @param {int} groupId The group id specifies which group name should be retrieved
 * @returns A string that holds the name of the group
 */
export const selectNameOfGroup = (state: GlobalModelState, groupId?: number): string => {
    if (groupId == null) {
        return '';
    }
    if (typeof state.models.modelGroups[groupId] !== 'undefined') {
        return state.models.modelGroups[groupId].name;
    }
    return '';
};
/**
 * This selector allows components to select the statistical values of a given group (specified by
 * ID)
 *
 * @category ModelSlice
 * @function
 * @param {object} state The global redux state
 * @param {int} groupId The group id specifies which data should be retrieved
 * @returns An object that maps each statistical value onto a boolean
 */
export const selectStatisticalValueSettingsOfGroup = (state: GlobalModelState, groupId: number) =>
    state.models.modelGroups[groupId].visibleSV;
/**
 * This selector allows components to select the visibility of a given group (specified by ID)
 *
 * @category ModelSlice
 * @function
 * @param {object} state The global redux state
 * @param {int} groupId The group id specifies which data should be retrieved
 * @returns Boolean value whether the group is visible or not
 */
export const selectVisibilityOfGroup = (state: GlobalModelState, groupId: number) =>
    state.models.modelGroups[groupId].isVisible;
/**
 * This selector allows components to select all valid group ids
 *
 * @category ModelSlice
 * @function
 * @param {object} state The global redux state
 * @returns An array holding all valid group ids
 */
export const selectAllGroupIds = (state: GlobalModelState) =>
    Object.keys(state.models.modelGroups).map((key) => parseInt(key));

/**
 * This selector allows components to select all Model Groups.
 *
 * @category ModelSlice
 * @function
 * @param {object} state The global redux state
 * @returns An array containing all Model Groups
 */
export const selectAllModelGroups = (state: GlobalModelState) => state.models.modelGroups;
