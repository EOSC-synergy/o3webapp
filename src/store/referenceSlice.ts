import { createSlice } from '@reduxjs/toolkit';
import { HYDRATE } from 'next-redux-wrapper';

/**
 * The string identifying the default reference model used in the webApp
 *
 * @category ReferenceSlice
 * @default 'SBUV_GSFC_merged-SAT-ozone'
 */
export const DEFAULT_REF_MODEL = 'SBUV_GSFC_merged-SAT-ozone';

/**
 * The default year of the reference year
 *
 * @category ReferenceSlice
 * @default 1980
 */
export const DEFAULT_REF_YEAR = 1980;

export type ReferenceState = {
    settings: {
        year: number;
        model: string;
        visible: boolean;
        isOffsetApplied: boolean;
    };
};
export type GlobalReferenceState = {
    reference: ReferenceState;
};

/**
 * The initial state of the referenceSlice defines the data structure in the store. Each reference
 * setting has its own settings i.e. title, ref Year etc.
 *
 * If you change this initial state you have to adapt the first test in the corresponding test file,
 * that tests the initial state.
 */
const initialState: ReferenceState = {
    settings: {
        year: DEFAULT_REF_YEAR,
        model: DEFAULT_REF_MODEL,
        visible: true,
        isOffsetApplied: false,
    },
};

type Payload<T> = {
    payload: T;
};

/**
 * The referenceSlice is generated by the redux toolkit. The reducers are defined here and the
 * corresponding actions are auto-generated.
 *
 * @category ReferenceSlice
 * @constant {Object}
 */
const referenceSlice = createSlice({
    name: 'reference',
    initialState,
    reducers: {
        /**
         * This reducer accepts an action object returned from setYear() e.g.
         * dispatch(setYear({year: 1980})) and calculates the new state based on the action and the
         * action data given in action.payload.
         *
         * In this case the current year is set to the given year.
         *
         * @example
         *     dispatch(setYear({ year: 1980 }));
         *
         * @param state The current store state of: state/reference
         * @param year The new year.
         */
        setYear(state: ReferenceState, { payload: { year } }: Payload<{ year: number }>) {
            state.settings.year = year;
        },

        /**
         * This reducer accepts an action object returned from setModel() e.g.
         * dispatch(setModel({model: "CCMI-1_CCCma_CMAM-refC2"})) and calculates the new state based
         * on the action and the action data given in action.payload.
         *
         * In this case the current model is set to the given model.
         *
         * @example
         *     dispatch(setModel({ model: 'CCMI-1_CCCma_CMAM-refC2' }));
         *
         * @param state The current store state of: state/reference
         * @param model The new reference model name.
         */
        setModel(state: ReferenceState, { payload: { model } }: Payload<{ model: string }>) {
            state.settings.model = model;
        },

        /**
         * This reducer accepts an action object returned from setVisibility() e.g.
         * dispatch(setVisibility({visible: true})) and calculates the new state based on the action
         * and the action data given in action.payload.
         *
         * In this case the current visibility is set to the given visibility.
         *
         * @example
         *     dispatch(setVisibility({ visible: true }));
         *
         * @param state The current store state of: state/reference
         * @param visible Whether the reference line should be visible.
         */
        setVisibility(
            state: ReferenceState,
            { payload: { visible } }: Payload<{ visible: boolean }>
        ) {
            state.settings.visible = visible;
        },

        /**
         * This reducer accepts an action object returned from setOffsetApplied() e.g.
         * dispatch(setOffsetApplied({isOffsetApplied: true})) and calculates the new state based on
         * the action and the action data given in action.payload.
         *
         * In this case the current offset status is set to the given offset status.
         *
         * @example
         *     dispatch(setOffsetApplied({ isOffsetApplied: true }));
         *
         * @param {object} state The current store state of: state/reference
         * @param {object} action Accepts the action returned from setOffsetApplied()
         * @param {object} action.payload The payload is an object containing the given data
         * @param {boolean} action.payload.isOffsetApplied The boolean for the offset status.
         */
        setOffsetApplied(
            state: ReferenceState,
            { payload: { isOffsetApplied } }: Payload<{ isOffsetApplied: boolean }>
        ) {
            state.settings.isOffsetApplied = isOffsetApplied;
        },
    },
    extraReducers: {
        [HYDRATE]: (state: ReferenceState, action) => {
            console.log('HYDRATE', state, action.payload);
            return {
                ...state,
                ...action.payload.subject,
            };
        },
    },
});

/**
 * The here listed actions are exported and serve as an interface for the view (our React
 * components).
 */
export const { setYear, setModel, setVisibility, setOffsetApplied } = referenceSlice.actions;

/**
 * The reducer combining all reducers defined in the reference slice. This has to be included in the
 * redux store, otherwise dispatching the above defined actions wouldn't trigger state updates.
 *
 * @category ReferenceSlice
 * @constant {Object}
 */
export default referenceSlice.reducer;

/**
 * This selector allows components to select the current reference year from the store.
 *
 * @category ReferenceSlice
 * @function
 * @param {object} state The global redux state
 * @returns {number} The current active reference year
 */
export const selectRefYear = (state: GlobalReferenceState) => state.reference.settings.year;

/**
 * This selector allows components to select the current reference model from the store.
 *
 * @category ReferenceSlice
 * @function
 * @param {object} state The global redux state
 * @returns {string} The current active reference model
 */
export const selectRefModel = (state: GlobalReferenceState) => state.reference.settings.model;

/**
 * This selector allows components to select the current visibility of the reference line from the
 * store.
 *
 * @category ReferenceSlice
 * @function
 * @param {object} state The global redux state
 * @returns {boolean} The current visibility of the reference line.
 */
export const selectVisibility = (state: GlobalReferenceState) => state.reference.settings.visible;

/**
 * This selector allows components to select the current status of the offset from the store.
 *
 * @category ReferenceSlice
 * @function
 * @param {object} state The global redux state
 * @returns {boolean} The current status of the offset.
 */
export const selectIsOffsetApplied = (state: GlobalReferenceState) =>
    state.reference.settings.isOffsetApplied;
