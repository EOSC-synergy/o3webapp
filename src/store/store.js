import { applyMiddleware, combineReducers, configureStore, createStore } from '@reduxjs/toolkit';
import plotReducer from './plotSlice/plotSlice';
import modelsReducer, { setModelsOfModelGroup } from './modelsSlice/modelsSlice';
import referenceReducer from './referenceSlice/referenceSlice';
import apiReducer from '../services/API/apiSlice/apiSlice';
import { DEFAULT_MODEL_GROUP } from '../utils/constants';
import { HYDRATE, createWrapper } from 'next-redux-wrapper';
import thunkMiddleware from 'redux-thunk';

const bindMiddleware = (middleware) => {
    if (process.env.NODE_ENV !== 'production') {
        const { composeWithDevTools } = require('redux-devtools-extension');
        return composeWithDevTools(applyMiddleware(...middleware));
    }
    return applyMiddleware(...middleware);
};

const combinedReducer = combineReducers({
    plot: plotReducer,
    models: modelsReducer,
    reference: referenceReducer,
    api: apiReducer,
});

const reducer = (state, action) => {
    if (action.type === HYDRATE) {
        const nextState = {
            ...state, // use previous state
            ...action.payload, // apply delta from hydration
        };
        // if (state.count.count) nextState.count.count = state.count.count; // preserve count value on client side navigation
        return nextState;
    } else {
        return combinedReducer(state, action);
    }
};

let store;

const initStore = () => {
    store = createStore(reducer, bindMiddleware([thunkMiddleware]));
    return store;
};

export { store };

export const wrapper = createWrapper(initStore);

export function createTestStore() {
    const store = configureStore({
        reducer: combinedReducer,
    });
    store.dispatch(setModelsOfModelGroup(DEFAULT_MODEL_GROUP)); // fill with default group
    return store;
}
