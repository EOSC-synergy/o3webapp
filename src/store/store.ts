import { Action, combineReducers, configureStore, ThunkAction } from '@reduxjs/toolkit';
import plotReducer, { GlobalPlotState } from './plotSlice';
import modelsReducer, { GlobalModelState, setModelsOfModelGroup } from './modelsSlice';
import referenceReducer, { GlobalReferenceState } from './referenceSlice';
import apiReducer from '../services/API/apiSlice/apiSlice';
import { DEFAULT_MODEL_GROUP } from '../utils/constants';
import { createWrapper, HYDRATE } from 'next-redux-wrapper';
import thunkMiddleware from 'redux-thunk';
import { useDispatch, useStore } from 'react-redux';

type GlobalState =
    | undefined
    | (GlobalPlotState & GlobalModelState & GlobalReferenceState & { api: any });

const combinedReducer = combineReducers({
    plot: plotReducer,
    models: modelsReducer,
    reference: referenceReducer,
    api: apiReducer,
});

const reducer = (state: GlobalState, action: { type: string; payload: any }) => {
    if (action.type === HYDRATE) {
        // if (state.count.count) nextState.count.count = state.count.count; // preserve count value on client side navigation
        return {
            ...state, // use previous state
            ...action.payload, // apply delta from hydration
        };
    } else {
        return combinedReducer(state, action);
    }
};

//let store: ReturnType<typeof configureStore<GlobalState, >>;

const initStore = () => {
    //store = createStore(reducer, bindMiddleware([thunkMiddleware]));
    return configureStore({
        reducer,
        devTools: process.env.NODE_ENV !== 'production',
        middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(thunkMiddleware),
    });
};

export type AppStore = ReturnType<typeof initStore>;
export type AppState = ReturnType<AppStore['getState']>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, AppState, unknown, Action>;

export type AppDispatch = AppStore['dispatch'];
export const useAppDispatch = () => useDispatch<AppDispatch>(); // Export a hook that can be reused to resolve types

export const useAppStore = () => useStore<AppState>();

export const wrapper = createWrapper(initStore);

export function createTestStore() {
    const store = initStore();
    store.dispatch(setModelsOfModelGroup(DEFAULT_MODEL_GROUP)); // fill with default group
    return store;
}
