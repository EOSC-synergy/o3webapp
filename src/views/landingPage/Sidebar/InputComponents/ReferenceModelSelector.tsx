import React, { type FC, useEffect } from 'react';
import { Autocomplete, TextField } from '@mui/material';
import { useSelector } from 'react-redux';
import { setModel, selectRefModel } from 'store/referenceSlice';
import { fetchPlotDataForCurrentModels, REQUEST_STATE } from 'services/API/apiSlice';
import { type AppState, useAppDispatch } from 'store';
import { type ErrorReporter } from 'utils/reportError';

type ReferenceModelSelectorProps = {
    reportError: ErrorReporter;
};
/**
 * Enables the user to select a reference model.
 * @returns a jsx containing a dropdown to select the reference model from all currently visible models
 */
const ReferenceModelSelector: FC<ReferenceModelSelectorProps> = ({ reportError }) => {
    /**
     * A dispatch function to dispatch actions to the Redux store.
     */
    const dispatch = useAppDispatch();

    /**
     * Selects the model list from the Redux store.
     * @constant {Array}
     */
    const modelListRequestedData = useSelector((state: AppState) => state.api.models);

    /**
     * Selects the current reference model.
     * @see {@link selectRefModel}
     */
    const selectedModel = useSelector(selectRefModel);

    /**
     * Array with all the models.
     */
    let allModels: string[] = [];
    if (modelListRequestedData.status === REQUEST_STATE.success) {
        allModels = modelListRequestedData.data;
    }

    useEffect(() => {
        if (modelListRequestedData.status === REQUEST_STATE.error) {
            reportError('API not responding: ' + modelListRequestedData.error);
        }
    });

    /**
     * Handles the change of the reference model selection.
     * @see {@link fetchPlotDataForCurrentModels}
     * @function
     */
    const handleChangeForRefModel = (_: unknown, value: string | null) => {
        if (value !== null) {
            dispatch(setModel({ model: value }));
            // fetch for tco3_zm and tco3_return
            dispatch(fetchPlotDataForCurrentModels());
        }
    };

    return (
        <Autocomplete
            id="locationAutocomplete"
            value={selectedModel}
            onChange={handleChangeForRefModel}
            options={allModels}
            renderInput={(params) => <TextField {...params} label="Reference Model" />}
            sx={{ width: '100%' }}
            data-testid="ReferenceModelSelector-reference-model"
        />
    );
};

export default ReferenceModelSelector;
