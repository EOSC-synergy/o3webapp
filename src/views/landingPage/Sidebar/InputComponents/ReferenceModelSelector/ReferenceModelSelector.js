import React, {useEffect} from "react";
import {Autocomplete, TextField} from "@mui/material";
import {useDispatch} from "react-redux";
import {useSelector} from "react-redux";
import {setModel, selectRefModel} from "../../../../../store/referenceSlice/referenceSlice";
import PropTypes from 'prop-types';
import {fetchPlotDataForCurrentModels} from "../../../../../services/API/apiSlice";
import { REQUEST_STATE } from "../../../../../services/API/apiSlice";
/**
 * enables the user to select a reference model
 * @component
 * @param {Object} props
 * @param {function} props.reportError - function to handle errors
 * @returns {JSX.Element} a jsx containing a dropdown to select the reference model from all currently visible models
 */
function ReferenceModelSelector(props) {

    /**
     * A dispatch function to dispatch actions to the Redux store.
     */
    const dispatch = useDispatch();

    /**
     * Selects the model list from the Redux store.
     */
    const modelListRequestedData = useSelector(state => state.api.models);

    /**
     * Selects the current reference model.
     */
    const selectedModel = useSelector(selectRefModel);

    /**
     * The Array with all the models.
     * @type {Array.<String>}
     */
    let allModels = [];
    if (modelListRequestedData.status === REQUEST_STATE.success) {
        allModels = modelListRequestedData.data;
    }

    useEffect(() => {
        if (modelListRequestedData.status === REQUEST_STATE.error) {
            props.reportError("API not responding: " + modelListRequestedData.error);
        }
    });

    /**
     * Handles the change of the reference model selection.
     */
    const handleChangeForRefModel = (event, value) => {
        if (value !== null) {
            dispatch(setModel({model: value}));
            // fetch for tco3_zm and tco3_return
            dispatch(fetchPlotDataForCurrentModels());
        }
    };

    return (
        <Autocomplete
            id="locationAutocomplete"
            value={selectedModel}
            onChange={handleChangeForRefModel}
            options={allModels}
            renderInput={(params) => <TextField {...params} label="Reference Model" />}
            sx={{width: "100%"}}
        />
    );
}

ReferenceModelSelector.propTypes = {
    reportError: PropTypes.func
}

export default ReferenceModelSelector;