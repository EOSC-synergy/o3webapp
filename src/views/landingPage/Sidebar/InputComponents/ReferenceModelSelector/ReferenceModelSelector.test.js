import React from 'react';
import ReferenceModelSelector from './ReferenceModelSelector';
import { render } from '@testing-library/react';
import "@testing-library/jest-dom/extend-expect";
import { createTestStore } from '../../../../../store/store';
import { Provider } from "react-redux";

let store;
beforeEach(() => {
  store = createTestStore();
});

describe("tests basic rendering", () => {
    it('Component renders without crashing', () => {
        render(<Provider store={store}><ReferenceModelSelector /></Provider>);
    
    });
    
    
    // Snapshot test
    it('renders correctly', () => {
        const { container } = render(<Provider store={store}><ReferenceModelSelector /></Provider>);
        expect(container).toMatchSnapshot();
    });
});

describe("tests redux functionality", () => {
    
});


