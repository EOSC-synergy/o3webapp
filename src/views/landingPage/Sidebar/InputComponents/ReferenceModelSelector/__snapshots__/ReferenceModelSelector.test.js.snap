// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    aria-expanded="false"
    class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon css-dd2h8b-MuiAutocomplete-root"
    role="combobox"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for="locationAutocomplete"
        id="locationAutocomplete-label"
      >
        Reference Model
      </label>
      <div
        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-autocomplete="list"
          aria-invalid="false"
          autocapitalize="none"
          autocomplete="off"
          class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
          id="locationAutocomplete"
          spellcheck="false"
          type="text"
          value="SBUV_GSFC_merged-SAT-ozone"
        />
        <div
          class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
        >
          <button
            aria-label="Clear"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
            tabindex="-1"
            title="Clear"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="Open"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
            tabindex="-1"
            title="Open"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-1z7n62"
          >
            <span>
              Reference Model
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;
